File: internal/src/main/java/dev/mvmo/sicklang/internal/env/SickEnvironment.java
package dev.mvmo.sicklang.internal.env;

import com.google.common.collect.Maps;
import dev.mvmo.sicklang.internal.object.SickObject;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

import java.util.Map;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class SickEnvironment {

    private final Map<String, SickObject> storageMap;
    private final SickEnvironment parent;

    public SickEnvironment set(String key, SickObject value) {
        this.storageMap.put(key, value);
        return this;
    }

    public SickObject setAndGet(String key, SickObject value) {
        set(key, value);
        return value;
    }

    public SickObject get(String key) {
        return storageMap.get(key);
    }

    public SickObject getOrDefault(String key, SickObject defaultValue) {
        if (hasKey(key))
            return get(key);
        return defaultValue;
    }

    public boolean hasKey(String key) {
        return storageMap.containsKey(key);
    }

    public static SickEnvironment newEnclosedInstance(SickEnvironment environment) {
        return new SickEnvironment(Maps.newHashMap(), environment);
    }

    public static SickEnvironment newInstance() {
        return new SickEnvironment(Maps.newHashMap(), null);
    }

}

File: internal/src/main/java/dev/mvmo/sicklang/internal/env/SickEnvironment.kt
package dev.mvmo.sicklang.internal.env

import dev.mvmo.sicklang.internal.`object`.SickObject

class SickEnvironment private constructor(
    private val storageMap: MutableMap<String, SickObject>,
    private val parent: SickEnvironment?
) {

    operator fun set(key: String, value: SickObject): SickEnvironment {
        storageMap[key] = value
        return this
    }

    fun setAndGet(key: String, value: SickObject): SickObject {
        set(key, value)
        return value
    }

    operator fun get(key: String): SickObject? {
        val `val`: SickObject? = storageMap[key]
        return if (`val` == null && parent != null) parent[key] else `val`
    }

    fun getOrDefault(key: String, defaultValue: SickObject?): SickObject? {
        return if (hasKey(key)) get(key) else defaultValue
    }

    fun hasKey(key: String): Boolean {
        val contains = storageMap.containsKey(key)
        return if (!contains && parent != null) parent.hasKey(key) else contains
    }

    companion object {
        @JvmStatic
        fun newEnclosedInstance(environment: SickEnvironment): SickEnvironment {
            return SickEnvironment(mutableMapOf(), environment)
        }

        @JvmStatic
        fun newInstance(): SickEnvironment {
            return SickEnvironment(mutableMapOf(), null)
        }
    }

}